sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ListStateImpl.scala

```diff
@@ -49,101 +50,187 @@ class ListStateImpl[S](
-   override def clear(): Unit = {
-     val encodedKey = stateTypesEncoder.encodeGroupingKey()
-     store.remove(encodedKey, stateName)
-     val entryCount = getEntryCount(encodedKey)
-     TWSMetricsUtils.incrementMetric(metrics, "numRemovedStateRows", entryCount)
-     removeEntryCount(encodedKey)
-   }
-
-   private def validateNewState(newState: Array[S]): Unit = {
-     StateStoreErrors.requireNonNullStateValue(newState, stateName)
-     StateStoreErrors.requireNonEmptyListStateValue(newState, stateName)
-
-     newState.foreach { v =>
-       StateStoreErrors.requireNonNullStateValue(v, stateName)
-     }
-   }
- }
+  override def exists(): Boolean = {
+    if (usingAvro) {
+      val encodedKey: Array[Byte] = avroTypesEncoder.encodeGroupingKey()
+      store.get(encodedKey, stateName) != null
+    } else {
+      val encodedKey = unsafeRowTypesEncoder.encodeGroupingKey()
+      store.get(encodedKey, stateName) != null
+    }
+  }
+
+  /**
+   * Get the state value if it exists. If the state does not exist in state store, an
+   * empty iterator is returned.
+   */
+  override def get(): Iterator[S] = {
+    if (usingAvro) {
+      getAvro()
+    } else {
+      getUnsafeRow()
+    }
+  }
+
+  private def getAvro(): Iterator[S] = {
+    val encodedKey: Array[Byte] = avroTypesEncoder.encodeGroupingKey()
+    val avroValuesIterator = store.valuesIterator(encodedKey, stateName)
+    new Iterator[S] {
+      override def hasNext: Boolean = {
+        avroValuesIterator.hasNext
+      }
+
+      override def next(): S = {
+        val valueRow = avroValuesIterator.next()
+        avroTypesEncoder.decodeValue(valueRow)
+      }
+    }
+  }
+
+  private def getUnsafeRow(): Iterator[S] = {
+    val encodedKey = unsafeRowTypesEncoder.encodeGroupingKey()
+    val unsafeRowValuesIterator = store.valuesIterator(encodedKey, stateName)
+    new Iterator[S] {
+      override def hasNext: Boolean = {
+        unsafeRowValuesIterator.hasNext
+      }
+
+      override def next(): S = {
+        val valueUnsafeRow = unsafeRowValuesIterator.next()
+        unsafeRowTypesEncoder.decodeValue(valueUnsafeRow)
+      }
+    }
+  }
+
+  /** Update the value of the list. */
+  override def put(newState: Array[S]): Unit = {
+    validateNewState(newState)
+
+    if (usingAvro) {
+      putAvro(newState)
+    } else {
+      putUnsafeRow(newState)
+    }
+  }
+
+  private def putAvro(newState: Array[S]): Unit = {
+    val encodedKey: Array[Byte] = avroTypesEncoder.encodeGroupingKey()
+    var isFirst = true
+    var entryCount = 0L
+    TWSMetricsUtils.resetMetric(metrics, "numUpdatedStateRows")
+
+    newState.foreach { v =>
+      val encodedValue = avroTypesEncoder.encodeValue(v)
+      if (isFirst) {
+        store.put(encodedKey, encodedValue, stateName)
+        isFirst = false
+      } else {
+        store.merge(encodedKey, encodedValue, stateName)
+      }
+      entryCount += 1
+      TWSMetricsUtils.incrementMetric(metrics, "numUpdatedStateRows")
+    }
```

Comment:
There is `updateEntryCount` in `putUnsafeRow` but not here. Why?

Reply:
Metrics aren't supported for byte array, will do in a follow-up.